File: leetcode/0001-1000/0872-leaf-similar-trees/LeafSimilarTrees.java
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;
import java.util.function.Function;

public class LeafSimilarTrees {
  public static void main(String[] args) {
    //
  }
}

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {
  }

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  public boolean leafSimilar(TreeNode lhs, TreeNode rhs) {
    Function<TreeNode, TreeNode[]> getLeaves = (root) -> {
      if (root == null) {
        return new TreeNode[] {};
      }

      List<TreeNode> leaves = new ArrayList<>();

      Deque<TreeNode> queue = new LinkedList<>();

      queue.offer(root);

      while (!queue.isEmpty()) {
        TreeNode node = queue.pollFirst();

        if (node.left == null && node.right == null) {
          leaves.add(node);
        }

        if (node.right != null) {
          queue.offerFirst(node.right);
        }

        if (node.left != null) {
          queue.offerFirst(node.left);
        }
      }

      return leaves.toArray(TreeNode[]::new);
    };

    TreeNode[] lefts = getLeaves.apply(lhs);
    TreeNode[] rights = getLeaves.apply(rhs);

    if (lefts.length != rights.length) {
      return false;
    }

    for (int i = 0; i < lefts.length; i++) {
      if (lefts[i].val != rights[i].val) {
        return false;
      }
    }
    //
    return true;
  }
}

File: leetcode/0001-1000/0872-leaf-similar-trees/LeafSimilarTrees.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

function leafSimilar(root1: TreeNode | null, root2: TreeNode | null): boolean {
  const getLeaves = (root: TreeNode | null) => {
    if (root == null) {
      return [];
    }

    const leaves: TreeNode[] = [];
    const queue: TreeNode[] = [root];

    while (queue.length > 0) {
      const node = queue.shift();

      if (node?.left == null && node?.right == null) {
        leaves.push(node!);
      }

      if (node?.right != null) {
        queue.unshift(node.right);
      }

      if (node?.left != null) {
        queue.unshift(node.left);
      }
    }

    return leaves;
  };

  const lefts = getLeaves(root1);
  const rights = getLeaves(root2);

  if (lefts.length !== rights.length) {
    return false;
  }

  return lefts.length === rights.length
    && lefts.every((element, index) => element.val === rights[index].val);
};
